database/models/comment.js

```diff
@@ -0,0 +1,197 @@
+	) as score`)
+
+/**
+ * @typedef Comment
+ * @type {object}
+ * @property {number} id - the comment id
+ * @property {string} text - the text of the comment
+ * @property {string} proposalHash - the proposal of the comment
+ * @property {number} createdBy - the user that created the comment
+ * @property {string} createdAt - the date the comment was created
+ * @property {string} updatedAt - the date the comment was last updated
+ */
+
+/**
+ * @typedef CommentAttributes
+ * @type {object}
+ * @property {string} text
+ * @property {string} proposalHash
+ * @property {number} createdBy
+ */
+
+/**
+ * @typedef CommentVote
+ * @type {object}
+ * @property {number} direction - the user's vote (1 = like, -1 = dislike, 0 = unvote)
+ * @property {number} createdBy - the user of the vote
+ * @property {string} createdAt - the date the vote was created
+ * @property {string} updatedAt - the date the vote was last updated
+ */
+
+/**
+ * @typedef CommentVoteAttributes
+ * @type {object}
+ * @property {number} direction - the user's vote (1 = like, -1 = dislike, 0 = unvote)
+ * @property {number} createdBy - the user of the vote
+ */
+
+/**
+ * @function create - creates a new comment
+ * @param {CommentAttributes} attrs - the attributes of the comment to create
+ * @param {*=} trx - the optional transaction context
+ * @return {Promise.<Comment>} - the newly created comment
+ */
+module.exports.create = async (attrs, trx) => {
+	const db = await getDbDriver()
+	const [comment] = await withTransaction(trx, db(table).insert(createTimestamps(attrs), '*'))
+	return comment
+}
+
+/**
+ * @function update - updates an existing comment
+ * @param {number} id - the id of the comment to update
+ * @param {CommentAttributes} attrs - the attributes of the comment to update
+ * @param {*=} trx - the optional transaction context
+ * @return {Promise.<Comment>} - the updated comment
+ */
+module.exports.update = async (id, attrs, trx) => {
+	const db = await getDbDriver()
+	const [comment] = await withTransaction(
+		trx,
+		db(table)
+			.where('id', id)
+			.update(updateTimestamps(attrs), '*')
+	)
+
+	return comment
+}
+
+/**
+ * @function - soft deletes a comment
+ * @param {number} id - the id of the comment to delete
+ * @param {*=} trx - the optional transaction context
+ * @return {Promise<?Comment>} - the updated comment
+ */
+module.exports.delete = async (id, trx) => {
+	const db = await getDbDriver()
+	const [comment] = await withTransaction(
+		trx,
+		db(table)
+			.where('id', id)
+			.update(deleteTimestamps(), '*')
+	)
+
+	return comment
+}
+
+/**
+ * @function findOne - finds an existing comment that matches the provided attributes
+ * @param {CommentAttributes} attrs - the attributes of the comment to match against
+ * @return {Promise.<?Comment>} - the matched comment, null if not found
+ */
+module.exports.findOne = async attrs => {
+	const db = await getDbDriver()
+	const comment = await db
+		.first('*', likes(db), dislikes(db), score(db))
```

Comment:
hmm, isn't joining tables more efficient?

Reply:
I think can be replaced with a join + group by. Will investigate.

Reply:
Seems like JOIN might be the better choice here: https://explainextended.com/2010/01/30/aggregates-subqueries-vs-group-by/

Especially since the 3 sub-queries are against the same table.

Reply:
For the future, we may even want to add a materialized view for better performance since I'm a bit concerned with joining to a the voting table when one comment could have hundreds of votes.